### PR TITLE
feat: Armor Transparency on Ghosts

### PR DESCRIPTION
**Pull request by YouSoPugly**
Added armor transparency for player ghosts

### CODE REVIEW

File: common/src/main/java/com/wynntils/mc/EventFactory.java [hunk 1]
```diff
@@ -10,64 +10,16 @@
...
 import com.wynntils.core.WynntilsMod;
-import com.wynntils.mc.event.AddEntityLookupEvent;
-import com.wynntils.mc.event.ArmSwingEvent;
-import com.wynntils.mc.event.BossHealthUpdateEvent;
-import com.wynntils.mc.event.ChatPacketReceivedEvent;
-import com.wynntils.mc.event.ChatSentEvent;
-import com.wynntils.mc.event.ChestMenuQuickMoveEvent;
-import com.wynntils.mc.event.ClientTickEvent;
-import com.wynntils.mc.event.CommandsPacketEvent;
+import com.wynntils.mc.event.*;
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/mc/EventFactory.java:13]: Do not use * imports, we don't allow it

File: common/src/main/java/com/wynntils/features/user/players/PlayerGhostArmorTransparencyFeature.java [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+import com.wynntils.core.features.properties.FeatureCategory;
+import com.wynntils.core.features.properties.FeatureInfo;
+import com.wynntils.mc.event.PlayerGhostArmorRenderEvent;
+import com.wynntils.wynn.utils.WynnPlayerUtils;
+import com.wynntils.wynn.utils.WynnUtils;
+import net.minecraft.world.entity.player.Player;
+import net.minecraftforge.eventbus.api.SubscribeEvent;
+
+@FeatureInfo(category = FeatureCategory.PLAYERS)
+public class PlayerGhostArmorTransparencyFeature extends UserFeature {
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/features/user/players/PlayerGhostArmorTransparencyFeature.java:15]: I get that you wanted to do this the "proper" way, but there is no need to make this a separate feature. Just add this is a config option in PlayerghostTransparency feature.

**Review by kristofbolyai — CHANGES_REQUESTED** [review]
Merge the two features, but the mixin looks good.

**Comment by kristofbolyai** [comment]
Actually I take that back. The event/mixin is not the proper way.

File: common/src/main/java/com/wynntils/mc/event/PlayerGhostArmorRenderEvent.java [hunk 3]
```diff
@@ -0,0 +1,29 @@
+package com.wynntils.mc.event;
+
+import com.wynntils.core.WynntilsMod;
+import net.minecraft.world.entity.LivingEntity;
+import net.minecraftforge.eventbus.api.Event;
+
+public class PlayerGhostArmorRenderEvent extends Event {
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/mc/event/PlayerGhostArmorRenderEvent.java:7]: Mark this cancelable

Comment (kristofbolyai) [reply common/src/main/java/com/wynntils/mc/event/PlayerGhostArmorRenderEvent.java:7]: `@Cancelable`

File: common/src/main/java/com/wynntils/mc/event/PlayerGhostArmorRenderEvent.java [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+
+    private LivingEntity entity;
+    private boolean renderGhostArmor;
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/mc/event/PlayerGhostArmorRenderEvent.java:10]: remove this

File: common/src/main/java/com/wynntils/mc/mixin/HumanoidArmorLayerMixin.java [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+import org.spongepowered.asm.mixin.injection.Inject;
+import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;
+
+@Mixin(HumanoidArmorLayer.class)
+public abstract class HumanoidArmorLayerMixin<T extends LivingEntity> {
+
+    @Inject(method = "render(Lcom/mojang/blaze3d/vertex/PoseStack;Lnet/minecraft/client/renderer/MultiBufferSource;ILnet/minecraft/world/entity/LivingEntity;FFFFFF)V", at = @At("INVOKE"), cancellable = true)
+    private void render(PoseStack matrixStack, MultiBufferSource buffer, int packedLight, T livingEntity, float limbSwing, float limbSwingAmount, float partialTicks, float ageInTicks, float netHeadYaw, float headPitch, CallbackInfo ci) {
+        PlayerGhostArmorRenderEvent event = EventFactory.onArmorRender(livingEntity);
+        if (!event.shouldRenderGhostArmor())
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/mc/mixin/HumanoidArmorLayerMixin.java:23]: event.isCancelled() then cancel

**Review by kristofbolyai — CHANGES_REQUESTED** [review]
This is how we do cancelable events.

File: common/src/main/java/com/wynntils/mc/mixin/HumanoidArmorLayerMixin.java [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+import org.spongepowered.asm.mixin.injection.At;
+import org.spongepowered.asm.mixin.injection.Inject;
+import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;
+
+@Mixin(HumanoidArmorLayer.class)
+public abstract class HumanoidArmorLayerMixin<T extends LivingEntity> {
+
+    @Inject(method = "render(Lcom/mojang/blaze3d/vertex/PoseStack;Lnet/minecraft/client/renderer/MultiBufferSource;ILnet/minecraft/world/entity/LivingEntity;FFFFFF)V", at = @At("INVOKE"), cancellable = true)
+    private void render(PoseStack matrixStack, MultiBufferSource buffer, int packedLight, T livingEntity, float limbSwing, float limbSwingAmount, float partialTicks, float ageInTicks, float netHeadYaw, float headPitch, CallbackInfo ci) {
+        PlayerGhostArmorRenderEvent event = EventFactory.onArmorRender(livingEntity);
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/mc/mixin/HumanoidArmorLayerMixin.java:22]: imo check if the entity is a player here, and only post the event if it is true

**Comment by Incompleteusern** [comment]
Didn't want to do this, thanks!

File: common/src/main/java/com/wynntils/mc/event/PlayerGhostArmorRenderEvent.java (hunk 4 above)
Comment (Incompleteusern) [reply common/src/main/java/com/wynntils/mc/event/PlayerGhostArmorRenderEvent.java:10]: I don't see why?

Comment (kristofbolyai) [reply common/src/main/java/com/wynntils/mc/event/PlayerGhostArmorRenderEvent.java:10]: we should use cancelable like everywhere else

Comment (kristofbolyai) [reply common/src/main/java/com/wynntils/mc/event/PlayerGhostArmorRenderEvent.java:10]: Also, can you rename the event PlayerArmorRenderEvent? That is more appropriate

**Pull request closed without merging**